clippy_lints/src/duration_suboptimal_units.rs

```diff
@@ -0,0 +1,181 @@
+use clippy_utils::msrvs::{self, Msrv};
+use clippy_utils::res::MaybeDef;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, QPath, RustcVersion};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::impl_lint_pass;
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Checks for instances where a `std::time::Duration` is constructed using a smaller time unit
+    /// when the value could be expressed more clearly using a larger unit.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Using a smaller unit for a duration that is evenly divisible by a larger unit reduces
+    /// readability. Readers have to mentally convert values, which can be error-prone and makes
+    /// the code less clear.
+    ///
+    /// ### Example
+    /// ```
+    /// use std::time::Duration;
+    ///
+    /// let dur = Duration::from_millis(5_000);
+    /// let dur = Duration::from_secs(180);
+    /// let dur = Duration::from_mins(10 * 60);
+    /// ```
+    ///
+    /// Use instead:
+    /// ```
+    /// use std::time::Duration;
+    ///
+    /// let dur = Duration::from_secs(5);
+    /// let dur = Duration::from_mins(3);
+    /// let dur = Duration::from_hours(10);
+    /// ```
+    #[clippy::version = "1.94.0"]
+    pub DURATION_SUBOPTIMAL_UNITS,
+    pedantic,
+    "constructing a `Duration` using a smaller unit when a larger unit would be more readable"
+}
+
+impl_lint_pass!(DurationSuboptimalUnits => [DURATION_SUBOPTIMAL_UNITS]);
+
+pub struct DurationSuboptimalUnits {
+    msrv: Msrv,
+}
+
+impl DurationSuboptimalUnits {
+    pub fn new(conf: &'static Conf) -> Self {
+        Self { msrv: conf.msrv }
+    }
+}
+
+impl LateLintPass<'_> for DurationSuboptimalUnits {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ Expr<'_>) {
+        // Check if a function on std::time::Duration is called
+        if let ExprKind::Call(func, [arg]) = expr.kind
+            && let ExprKind::Path(QPath::TypeRelative(func_ty, func_name)) = func.kind
+            && cx
+                .typeck_results()
+                .node_type(func_ty.hir_id)
+                .is_diag_item(cx, sym::Duration)
+        {
+            // We intentionally don't want to evaluate referenced constants, as we don't want to
+            // recommend a literal value over using constants:
+            //
+            // let dur = Duration::from_secs(SIXTY);
+            //           ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Duration::from_mins(1)`
+            let Some(Constant::Int(value)) = ConstEvalCtxt::new(cx).eval_local(arg, arg.span.ctxt()) else {
+                return;
+            };
+            // There is no need to promote e.g. 0 seconds to 0 hours
+            if value == 0 {
+                return;
+            }
+
+            let function_name = func_name.ident.name.as_str();
+            let Some((promoted_function_name, promoted_value)) = self.promote(cx, function_name, value) else {
```

Comment:
All values for those functions fit in a `u64`:


```suggestion
            let Some((promoted_function_name, promoted_value)) =
                self.promote(cx, func_name.ident.name, u64::try_from(value).unwrap())
            else {
```

Reply:
I mistakenly though stuff like `from_nanos` takes an u128. But all constructors we are using here take an u64 and only lower the value, so this is fine (tm).
Updated the code (i used `expect` instead of `unwrap`, hope that's ok)